### ISSUE TITLE
prefix does not filter company candidates

### ISSUE THREAD

**Issue by ztlevi** [emacs-lsp/lsp-javascript#9]
I'm trying to add `compnay-lsp` backend to `company-backends-js2-mode`. But once I add the `company-lsp` backend, company mode does not working properly to filter candidates when I typing.

This is how it looks like:

![791219c27c491f055320fee8f6c5bd034a7ffd75](https://user-images.githubusercontent.com/16655096/38182157-6c365f42-35ec-11e8-9189-ab99d2fc018e.gif)

Here is my setting:
```emacs lisp
(use-package lsp-javascript-typescript
    :commands (lsp-javascript-typescript-enable)
    :init
    (add-hook 'js-mode-hook #'lsp-javascript-typescript-enable)
    (add-hook 'typescript-mode-hook #'lsp-javascript-typescript-enable) ;; for typescript support
    (add-hook 'js2-mode-hook #'lsp-javascript-typescript-enable) ;; for js2-mode support
    (add-hook 'rjsx-mode #'lsp-javascript-typescript-enable) ;; for rjsx-mode support
    (spacemacs|add-company-backends
      :backends company-lsp
      :modes js2-mode)
    :defer t)
```

The value for `company-backends-js2-mode` is: 
```
Its value is ((company-lsp :with company-yasnippet)
 (company-dabbrev-code company-gtags company-etags company-keywords :with company-yasnippet)
 (company-files :with company-yasnippet)
 (company-dabbrev :with company-yasnippet))
```

I only have this issue with `lsp-javascript`. It works for lsp-python and cquery. So I assume it's `lsp-javascript` bug.

**Comment by gpittarelli** [comment]
Hmm, lsp-javascript is just a very light config telling lsp-mode how to work for javascript files; there's basically no relevant logic unique to lsp-javascript vs python or cquery that could cause; perhaps its specific to the typescript language server

**Comment by ztlevi** [comment]
Yeah. But I'm also confused that we should enable all the supports or just enable the one we need? Do they have conflicts with each other?

**Comment by sebastiencs** [comment]
Yes `javascript-typescript-langserver` doesn't take into account the prefix.
I'm using this code to fix it:

```el
(defun my-company-transformer (candidates)
  (let ((completion-ignore-case t))
    (all-completions (company-grab-symbol) candidates)))

(defun my-js-hook nil
  (make-local-variable 'company-transformers)
  (push 'my-company-transformer company-transformers))

(add-hook 'js-mode-hook 'my-js-hook)
```

**Comment by gpittarelli** [comment]
Nice fix. I'd like to find some way to work this into lsp-javascript-typescript directly (and figure out if we need similar fixes for other completion frontends). For now I've added it to the README.

**Comment by jvalkeejarvi** [comment]
> Yes `javascript-typescript-langserver` doesn't take into account the prefix.
> I'm using this code to fix it:
> 
> ```emacs-lisp
> (defun my-company-transformer (candidates)
>   (let ((completion-ignore-case t))
>     (all-completions (company-grab-symbol) candidates)))
> 
> (defun my-js-hook nil
>   (make-local-variable 'company-transformers)
>   (push 'my-company-transformer company-transformers))
> 
> (add-hook 'js-mode-hook 'my-js-hook)
> ```
> 
> You will get only condidates that start with the prefix.

Adding a transformer like this breaks company-file so file name completions don't work correctly. As a workaround I have edited the transformer function to do the transformation only when completion candidates are returned by lsp. It's not the most elegant solution but it works.

```emacs-lisp
  (defun my-company-transformer (candidates)
    (let ((completion-ignore-case t))
      (if (and (car candidates)
                   (get-text-property 0 'lsp-completion-prefix (car candidates)))
          (all-completions (company-grab-symbol) candidates)
        candidates
        )
      )
    )

```
